numba/npyufunc/parallel.py

```diff
@@ -35,58 +38,48 @@ def get_thread_count():
 
 
-class ParallelUFuncBuilder(ufuncbuilder.UFuncBuilder):
-    def build(self, cres, sig):
-        _launch_threads()
-        _init()
-
-        # Buider wrapper for ufunc entry point
-        ctx = cres.target_context
-        signature = cres.signature
-        library = cres.library
-        fname = cres.fndesc.llvm_func_name
-
-        ptr = build_ufunc_wrapper(library, ctx, fname, signature, cres)
-        # Get dtypes
-        dtypenums = [np.dtype(a.name).num for a in signature.args]
-        dtypenums.append(np.dtype(signature.return_type.name).num)
-        keepalive = ()
-        return dtypenums, ptr, keepalive
-
-
-def build_ufunc_wrapper(library, ctx, fname, signature, cres):
-    innerfunc = ufuncbuilder.build_ufunc_wrapper(library, ctx, fname,
-                                                 signature, objmode=False,
-                                                 cres=cres)
-    return build_ufunc_kernel(library, ctx, innerfunc, signature)
-
-
-def build_ufunc_kernel(library, ctx, innerfunc, sig):
-    """Wrap the original CPU ufunc with a parallel dispatcher.
+def build_gufunc_kernel(library, ctx, innerfunc, sig, inner_ndim):
+    """Wrap the original CPU ufunc/gufunc with a parallel dispatcher.
+    This function will wrap gufuncs and ufuncs something like.
 
     Args
     ----
     ctx
         numba's codegen context
 
     innerfunc
-        llvm function of the original CPU ufunc
+        llvm function of the original CPU gufunc
 
     sig
-        type signature of the ufunc
+        type signature of the gufunc
+
+    inner_ndim
+        inner dimension of the gufunc (this is len(sig.args) in the case of a
+        ufunc)
 
     Details
     -------
 
-    Generate a function of the following signature:
+    The kernel signature looks like this:
 
-    void ufunc_kernel(char **args, npy_intp *dimensions, npy_intp* steps,
-                      void* data)
+    void kernel(char **args, npy_intp *dimensions, npy_intp* steps, void* data)
 
-    Divide the work equally across all threads and let the last thread take all
-    the left over.
+    args - the input arrays + output arrays
+    dimensions - the dimensions of the arrays
+    steps - the step size for the array (this is like sizeof(type))
+    data - any additional data
 
+    The parallel backend then stages multiple calls to this kernel concurrently
+    across a number of threads. Practically, for each item of work, the backend
+    duplicates `dimensions` and adjusts the first entry to reflect the size of
+    the item of work, it also forms up an array of pointers into the args for
+    offsets to read/write from/to with respect to its position in the items of
+    work. This allows the same kernel to be used for each item of work, with
+    simply adjusted reads/writes/domain sizes and is safe by virtue of the
+    domain partitioning.
 
+    NOTE: The execution backend is passed the requested thread count, but it can
+    choose to ignore it (TBB)!
```

Comment:
let's probably just remove it? the only affected implementation is workqueue, and there is a way to store the chosen number of threads